rllib/agents/mbmpo/mbmpo.py

```diff
@@ -0,0 +1,328 @@
+
+            res = self.metric_gen.__call__(None)
+            res.update(self.metrics)
+            self.step_counter = 0
+            print("MB-MPO Iteration Completed")
+            return [res]
+        else:
+            print("MAML Iteration {} Completed".format(self.step_counter))
+            self.step_counter += 1
+
+            # Sync workers with meta policy
+            self.workers.sync_weights()
+            return []
+
+    def postprocess_metrics(self, metrics, prefix=""):
+        for key in metrics.keys():
+            self.metrics[prefix + "_" + key] = metrics[key]
+
+
+def post_process_metrics(prefix, workers, metrics):
+    # Obtain Current Dataset Metrics and filter out
+    res = collect_metrics(remote_workers=workers.remote_workers())
+    for key in METRICS_KEYS:
+        metrics[prefix + "_" + key] = res[key]
+    return metrics
+
+
+def inner_adaptation(workers, samples):
+    # Each worker performs one gradient descent
+    for i, e in enumerate(workers.remote_workers()):
+        e.learn_on_batch.remote(samples[i])
+
+
+def fit_dynamics(policy, pid):
+    return policy.dynamics_model.fit()
+
+
+def sync_ensemble(workers, model_attr="dynamics_model"):
+    def get_ensemble_weights(worker):
+        policy_map = worker.policy_map
+        policies = policy_map.keys()
+
+        def policy_ensemble_weights(policy):
+            model = policy.dynamics_model
+            return {
+                k: v.cpu().detach().numpy()
+                for k, v in model.state_dict().items()
+            }
+
+        return {
+            pid: policy_ensemble_weights(policy)
+            for pid, policy in policy_map.items() if pid in policies
+        }
+
+    def set_ensemble_weights(policy, pid, weights):
+        weights = weights[pid]
+        weights = convert_to_torch_tensor(weights, device=policy.device)
+        model = policy.dynamics_model
+        model.load_state_dict(weights)
+
+    if workers.remote_workers():
+        weights = ray.put(get_ensemble_weights(workers.local_worker()))
+        set_func = ray.put(set_ensemble_weights)
+        for e in workers.remote_workers():
+            e.foreach_policy.remote(set_func, weights=weights)
+
+
+def sync_stats(workers):
+    def get_normalizations(worker):
+        policy = worker.policy_map[DEFAULT_POLICY_ID]
+        return policy.dynamics_model.normalizations
+
+    def set_normalizations(policy, pid, normalizations):
+        policy.dynamics_model.set_norms(normalizations)
+
+    if workers.remote_workers():
+        normalization_dict = ray.put(
+            get_normalizations(workers.local_worker()))
+        set_func = ray.put(set_normalizations)
+        for e in workers.remote_workers():
+            e.foreach_policy.remote(
+                set_func, normalizations=normalization_dict)
+
+
+# Similar to MAML Execution Plan
+def execution_plan(workers, config):
+    # Train TD Models
+    workers.local_worker().foreach_policy(fit_dynamics)[0]
```

Comment:
why do we need the [0] index here? It's not assigned to anything.

Reply:
Foreach_policy returns a list but yeah dont need it